File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        maxPathOverall = INT_MIN;
        dfs(root);

        return maxPathOverall;
    }

private:
    int maxPathOverall;

    int dfs(TreeNode* root) {
        if (root == nullptr)
            return 0;

        int leftGain = max(dfs(root->left), 0);
        int rightGain = max(dfs(root->right), 0);

        maxPathOverall = max(root->val + leftGain + rightGain, maxPathOverall);

        return root->val + max(leftGain, rightGain);
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-4.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        return dfs(root, root.val);
    }

    public int dfs(TreeNode root, int maxSoFar) {
        if (root == null)
            return 0;

        int ans = dfs(root.left, Math.max(maxSoFar, root.val)) + dfs(root.right, Math.max(maxSoFar, root.val));

        if (root.val >= maxSoFar)
            ++ans;

        return ans;
    }
}
